File: python/problem-14.py
def collatz(n):
    while n != 1:
        yield n
        if n % 2:
            m = 3*n + 1
        else:
            m = n/2
        n = m
    else:
        yield 1

def largest_collatz(n):
    return max((sum(1 for n in collatz(x)),x) for x in range(n,1,-1))

largest_collatz(1000000) # => (525, 837799)

File: clojure/problem_14.clj
(defn collatz-step [n]
  (if (even? n)
    (/ n 2)
    (+ 1 (* 3 n))))

(defn collatz [start]
  (take-while #(not= % 1) (iterate collatz-step start)))

(defn collatz-len [start] (count (collatz start)))

(def result
  (second
   (apply max-key first
          (map
           #(identity [(collatz-len %) %])
           (range 2 1000000)))))

(println result)
